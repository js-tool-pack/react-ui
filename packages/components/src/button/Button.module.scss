@import '../scss.variable';

:global {
  .#{$PREFIX}button {
    --t-btn-main-color: black;
    --t-btn-color: white;
    --t-btn-bg-color: unset;
    --t-btn-border-style: solid;
    --t-btn-border-color: unset;
    --t-btn-border-width: 0;
    --t-btn-radius: var(--t-redis);
    --t-btn-padding: 5px;

    --t-btn-hover-color: white;
    --t-btn-hover-border-color: var(--t-btn-hover-color);
    --t-btn-hover-bg-color: unset;

    --t-btn-danger-color: red;
    --t-btn-danger-hover-color: #ff7875;

    position: relative;
    border-radius: var(--t-btn-radius);
    padding: 0 var(--t-btn-padding);
    height: var(--t-size);
    outline: 0;
    cursor: pointer;
    border: var(--t-btn-border-width) var(--t-btn-border-style)
      var(--t-btn-border-color);
    background-color: var(--t-btn-bg-color);
    color: var(--t-btn-color);
    transition: all var(--t-transition-duration) ease-in-out;

    &:not([disabled]):hover {
      --t-btn-color: var(--t-btn-hover-color);
      --t-btn-bg-color: var(--t-btn-hover-bg-color);
      --t-btn-border-color: var(--t-btn-hover-border-color);
    }

    &[disabled] {
      cursor: not-allowed;
    }
    &.type-primary {
      --t-btn-main-color: var(--t-primary-bg-color);
      --t-btn-bg-color: var(--t-primary-bg-color);
      --t-btn-hover-bg-color: var(--t-primary-hover-bg-color);
      &[disabled] {
        --t-btn-bg-color: var(--t-primary-disabled-bg-color);
      }
    }
    &.type-success {
      --t-btn-main-color: #67c23a;
      --t-btn-bg-color: #67c23a;
      --t-btn-hover-bg-color: #a9ea89;
      &[disabled] {
        --t-btn-bg-color: #c5f6ac;
      }
    }
    &.type-warning {
      --t-btn-main-color: #e6a23c;
      --t-btn-bg-color: #e6a23c;
      --t-btn-hover-bg-color: #eebe77;
      &[disabled] {
        --t-btn-bg-color: #f3cf9a;
      }
    }
    &.type-info {
      --t-btn-main-color: #909399;
      --t-btn-bg-color: #909399;
      --t-btn-hover-bg-color: #b1b3b8;
      &[disabled] {
        --t-btn-bg-color: #d5d7da;
      }
    }
    &.type-danger {
      --t-btn-main-color: #f56c6c;
      --t-btn-bg-color: #f56c6c;
      --t-btn-hover-bg-color: #f89898;
      &[disabled] {
        --t-btn-bg-color: #f89898;
      }
    }

    &.plain {
      --t-btn-border-width: 1px;
      --t-btn-border-style: solid;
      --t-btn-border-color: var(--t-btn-bg-color);
      --t-btn-color: var(--t-btn-bg-color);

      --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
      --t-btn-hover-color: var(--t-btn-hover-bg-color);
      background-color: unset;
    }

    &.plain-dashed {
      --t-btn-border-width: 1px;
      --t-btn-border-style: dashed;
      --t-btn-border-color: var(--t-btn-bg-color);
      --t-btn-color: var(--t-btn-bg-color);

      --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
      --t-btn-hover-color: var(--t-btn-hover-bg-color);
      background-color: unset;
    }
    &.plain-text {
      --t-btn-border-width: 0;
      --t-btn-border-style: dashed;
      --t-btn-border-color: var(--t-btn-bg-color);
      --t-btn-color: var(--t-btn-bg-color);

      --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
      --t-btn-hover-color: var(--t-btn-hover-bg-color);
      background-color: unset;

      &:not([disabled]):active {
        transition: all calc(var(--t-transition-duration) / 3) ease-out;
        transform: translateY(-5px);
      }
    }

    &.type-default {
      --t-btn-main-color: var(--t-primary-bg-color);
      --t-btn-border-width: 1px;
      --t-btn-color: var(--t-text-color);
      --t-btn-border-color: black;
      --t-btn-hover-border-color: var(--t-primary-bg-color);
      --t-btn-hover-color: var(--t-primary-bg-color);
      &[disabled] {
        --t-btn-color: #cecece;
        --t-btn-border-color: #cecece;
      }
      &.plain-text {
        --t-btn-border-width: 0;
      }
    }
    &.#{$PREFIX}size-lg {
      --t-btn-padding: 10px;
    }
    &.shape-round,
    &.shape-circle {
      --t-btn-radius: calc(var(--t-size) / 2);
    }
    &.shape-circle {
      width: var(--t-size);
      //overflow: hidden;
      //text-overflow: ellipsis;
      //white-space: nowrap;
      &.#{$PREFIX}size-lg {
        --t-btn-padding: 5px;
      }
    }

    .btn-wave {
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      border-radius: inherit;
      animation-duration: 0.5s;
      animation-iteration-count: 1;
    }
    .btn-wave-active {
      animation-name: btn-wave;
    }
  }

  @keyframes :global(btn-wave) {
    0% {
      box-shadow: 0 0 0.5px 0 var(--t-btn-main-color);
      opacity: 0.65;
    }

    100% {
      box-shadow: 0 0 0.5px 4.5px var(--t-btn-main-color);
      opacity: 0;
    }
  }
}
