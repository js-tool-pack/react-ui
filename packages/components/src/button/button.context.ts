import { createContext } from 'react';
import type { ButtonProps } from './button.types';

export const ButtonContext = createContext<ButtonProps>({});
