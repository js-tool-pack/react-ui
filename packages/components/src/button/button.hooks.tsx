import React, { useEffect, useMemo, useRef, useState } from 'react';
import { getClassNames } from '@tool-pack/basic';
import { CollapseTransition } from '../collapse-transition';
import { Icon } from '../icon';
import { Loading as LoadingIcon } from '@pkg/icons';
import type { ButtonProps } from './button.types';

export function useBtnWave(
  rootClass: string,
): readonly [boolean | React.ReactElement, () => void] {
  const [active, setActive] = useState(false);
  const waveRef = useRef<HTMLElement>(null);
  const activationTime = useRef(0);

  const activateWave = () => {
    if (active) {
      if (Date.now() - activationTime.current > 1500) setActive(false);
      return;
    }
    activationTime.current = Date.now();
    setActive(true);
  };

  useEffect(() => {
    const el = waveRef.current;
    if (!active || !el) return;
    const cancel = () => {
      el.removeEventListener('animationend', handler);
      el.removeEventListener('animationcancel', handler);
    };
    const handler = () => {
      setActive(false);
      cancel();
    };

    if (active && el) {
      el.addEventListener('animationend', handler);
      el.addEventListener('animationcancel', handler);
    }

    return handler;
  }, [active]);

  const wave = active && (
    <span
      ref={waveRef}
      className={getClassNames(`${rootClass}__wave`, {
        [`${rootClass}__wave--active`]: active,
      })}></span>
  );

  return [wave, activateWave];
}

export function useBtnIcon(
  rootClass: string,
  icon: ButtonProps['icon'],
  loading: ButtonProps['loading'],
): React.ReactElement {
  return useMemo(() => {
    const className = `${rootClass}__icon`;
    const Loading = (
      <Icon className={className} key="loading-icon">
        <LoadingIcon />
      </Icon>
    );

    if (icon)
      return loading ? Loading : <Icon className={className}>{icon}</Icon>;
    return <CollapseTransition width>{loading && Loading}</CollapseTransition>;
  }, [loading, icon]);
}
