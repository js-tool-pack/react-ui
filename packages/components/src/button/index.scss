@use '../namespace' as Name;
@use '../scss.variable' as Var;

$ct: Name.$collapse-transition;
$r: Name.$button;
:root {
  --t-btn-wave-opcity: 0.6;
  &.dark {
    --t-btn-wave-opcity: 1;
  }
}

.#{$r} {
  --t-btn-main-color: var(--t-text-color);
  --t-btn-color: white;
  --t-btn-bg-color: unset;
  --t-btn-border-style: solid;
  --t-btn-border-color: unset;
  --t-btn-border-width: 0;
  --t-btn-radius: var(--t-radius);
  --t-btn-padding: 12px;
  --t-btn-icon-space: 6px;

  // ------
  --t-btn-hover-color: white;
  --t-btn-hover-border-color: var(--t-btn-hover-color);
  --t-btn-hover-bg-color: unset;

  // ------
  --t-btn-danger-color: red;
  --t-btn-danger-hover-color: #ff7875;

  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  padding: 0 var(--t-btn-padding);
  height: var(--t-size);
  font-size: 1em;
  font-weight: 400;
  border: var(--t-btn-border-width) var(--t-btn-border-style)
    var(--t-btn-border-color);
  border-radius: var(--t-btn-radius);
  color: var(--t-btn-color);
  background-color: var(--t-btn-bg-color);
  outline: 0;
  user-select: none;
  cursor: pointer;
  transition-property: background-color, width, height, color, padding-top,
    padding-left, padding-right, padding-bottom, opacity;

  // transition-property: all;
  transition-timing-function: ease-in-out;
  transition-duration: var(--t-transition-duration);
  & > span {
    display: inline-flex;
    align-items: center;
  }
  &:not([disabled]):hover {
    --t-btn-color: var(--t-btn-hover-color);
    --t-btn-bg-color: var(--t-btn-hover-bg-color);
    --t-btn-border-color: var(--t-btn-hover-border-color);
  }
  &[disabled] {
    cursor: not-allowed;
  }
  &.#{$r}--type-primary {
    --t-btn-main-color: var(--t-primary-bg-color);
    --t-btn-bg-color: var(--t-primary-bg-color);
    --t-btn-hover-bg-color: var(--t-primary-hover-bg-color);
    &[disabled] {
      --t-btn-bg-color: var(--t-primary-disabled-bg-color);
    }
  }
  &.#{$r}--type-success {
    --t-btn-main-color: #67c23a;
    --t-btn-bg-color: #67c23a;
    --t-btn-hover-bg-color: #a9ea89;
    &[disabled] {
      --t-btn-bg-color: #c5f6ac;
    }
  }
  &.#{$r}--type-warning {
    --t-btn-main-color: #e6a23c;
    --t-btn-bg-color: #e6a23c;
    --t-btn-hover-bg-color: #eebe77;
    &[disabled] {
      --t-btn-bg-color: #f3cf9a;
    }
  }
  &.#{$r}--type-info {
    --t-btn-main-color: #909399;
    --t-btn-bg-color: #909399;
    --t-btn-hover-bg-color: #b1b3b8;
    &[disabled] {
      --t-btn-bg-color: #d5d7da;
    }
  }
  &.#{$r}--type-danger {
    --t-btn-main-color: #f56c6c;
    --t-btn-bg-color: #f56c6c;
    --t-btn-hover-bg-color: #f89898;
    &[disabled] {
      --t-btn-bg-color: #f89898;
    }
  }
  &.#{$r}--plain {
    --t-btn-border-width: 1px;
    --t-btn-border-style: solid;
    --t-btn-border-color: var(--t-btn-bg-color);
    --t-btn-color: var(--t-btn-bg-color);
    --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
    --t-btn-hover-color: var(--t-btn-hover-bg-color);

    background-color: unset;
  }
  &.#{$r}--plain-dashed {
    --t-btn-border-width: 1px;
    --t-btn-border-style: dashed;
    --t-btn-border-color: var(--t-btn-bg-color);
    --t-btn-color: var(--t-btn-bg-color);
    --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
    --t-btn-hover-color: var(--t-btn-hover-bg-color);

    background-color: unset;
  }
  &.#{$r}--plain-text {
    --t-btn-border-width: 0;
    --t-btn-border-style: dashed;
    --t-btn-border-color: var(--t-btn-bg-color);
    --t-btn-color: var(--t-btn-bg-color);
    --t-btn-hover-border-color: var(--t-btn-hover-bg-color);
    --t-btn-hover-color: var(--t-btn-hover-bg-color);

    background-color: unset;
    &:not([disabled]):active > * {
      transition: transform calc(var(--t-transition-duration) / 3) ease-out;
      transform: translateY(-5px);
    }
  }
  &.#{$r}--type-default {
    --t-btn-main-color: var(--t-primary-bg-color);
    --t-btn-border-width: 1px;
    --t-btn-color: var(--t-text-color);
    --t-btn-border-color: var(--t-btn-color);
    --t-btn-hover-border-color: var(--t-primary-bg-color);
    --t-btn-hover-color: var(--t-primary-bg-color);
    &[disabled] {
      --t-btn-color: #cecece;
      --t-btn-border-color: #cecece;
    }
    &.#{$r}--plain-text {
      --t-btn-border-width: 0;
    }
  }
  &.#{Var.$class-size-lg} {
    --t-btn-padding: 16px;
    --t-btn-icon-space: 8px;

    font-size: 16px;
  }
  &.#{Var.$class-size-sm} {
    --t-btn-padding: 6px;
    --t-btn-icon-space: 3px;
  }
  &.#{$r}--shape-round,
  &.#{$r}--shape-circle {
    --t-btn-radius: calc(var(--t-size) / 2);
  }
  &.#{$r}--shape-circle {
    width: var(--t-size);

    // overflow: hidden;
    // text-overflow: ellipsis;
    // white-space: nowrap;
    &.#{Var.$class-size-lg} {
      --t-btn-padding: 5px;
    }
  }
  &.#{$r}--shape-none {
    --t-btn-radius: 0;
  }
  .#{$r}__wave {
    position: absolute;
    inset: 0;
    border-radius: inherit;
    animation-duration: 0.5s;
    animation-iteration-count: 1;
  }

  .#{$r}__wave--active {
    animation-name: btn-wave;
  }
  &--icon-l {
    &:not(.#{$r}--icon-only) > .#{$r}__icon {
      margin-inline-end: var(--t-btn-icon-space);
    }
  }
  &--icon-r {
    &:not(.#{$r}--icon-only) > .#{$r}__icon {
      margin-inline-start: var(--t-btn-icon-space);
    }
  }
  &__icon {
    &.#{$ct}-enter-from,
    &.#{$ct}-leave-to {
      opacity: 0;
    }
  }
  &.#{$r}--loading {
    cursor: wait;
    opacity: 0.75;
  }
}

@keyframes btn-wave {
  0% {
    opacity: var(--t-btn-wave-opcity);
    box-shadow: 0 0 0.5px 0 var(--t-btn-main-color);
  }
  100% {
    opacity: 0;
    box-shadow: 0 0 0.5px 5px var(--t-btn-main-color);
  }
}
