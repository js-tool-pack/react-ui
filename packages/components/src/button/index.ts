export * from './button.types';
export * from './Button';
export * from './ButtonGroup';
export * from './button.context';
