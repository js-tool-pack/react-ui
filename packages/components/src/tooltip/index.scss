@use '../namespace' as Name;
@use '../popover/index' as Pop;

$r: Name.$tooltip;

.#{$r} {
  --t-tootip-transition-duration: calc(var(--t-transition-duration) / 2);

  @include Pop.pop-style(var(--t-tootip-transition-duration));
}
