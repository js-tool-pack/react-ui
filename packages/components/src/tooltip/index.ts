export type { TooltipProps } from './tooltip.types';
export * from './Tooltip';
