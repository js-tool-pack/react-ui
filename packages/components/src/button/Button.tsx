import React, { useEffect, useRef, useState } from 'react';
import { ButtonProps } from './button.types';
import {
  CLASS_SIZE_LG,
  CLASS_SIZE_M,
  CLASS_SIZE_SM,
  getComponentClass,
} from '@pkg/shared';
import { getClassNames } from '@tool-pack/basic';
import './Button.module.scss';

const cClass = getComponentClass('button');

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      className,
      htmlType,
      onClick,
      plain,
      size,
      type,
      children,
      shape,
      ...rest
    } = props;

    const [active, setActive] = useState(false);
    const waveRef = useRef<HTMLElement>(null);

    useEffect(() => {
      const el = waveRef.current;
      if (!active || !el) return;
      const cancel = () => el.removeEventListener('animationend', handler);
      const handler = () => {
        setActive(false);
        cancel();
      };
      el.addEventListener('animationend', handler);

      return cancel;
    }, [active]);

    const clickHandler: React.MouseEventHandler<HTMLButtonElement> = (e) => {
      if (props.disabled) return;
      if (plain !== 'text') setActive(true);
      return onClick?.(e);
    };

    return (
      <button
        {...rest}
        ref={ref}
        type={htmlType}
        onClick={clickHandler}
        className={getClassNames(cClass, `type-${type}`, {
          [CLASS_SIZE_SM]: size === 'small',
          [CLASS_SIZE_M]: size === 'medium',
          [CLASS_SIZE_LG]: size === 'large',
          [`shape-round`]: shape === 'round',
          [`shape-circle`]: shape === 'circle',
          [`plain`]: plain === true,
          [`plain-text`]: plain === 'text',
          [`plain-dashed`]: plain === 'dashed',
          [className as string]: className,
        })}>
        {children}
        <span
          ref={waveRef}
          className={getClassNames('btn-wave', {
            'btn-wave-active': active,
          })}></span>
      </button>
    );
  },
);

Button.displayName = 'Button';
Button.defaultProps = {
  size: 'medium',
  type: 'default',
};
