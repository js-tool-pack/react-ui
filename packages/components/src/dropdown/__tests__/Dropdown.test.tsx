import { fireEvent, render, act } from '@testing-library/react';
import { DropdownOptionsItem, Dropdown } from '..';
import { nextTick } from '@tool-pack/basic';
import { Tooltip } from '~/tooltip';
import { Button } from '~/button';

describe('Dropdown', () => {
  // 模拟 ResizeObserver，ResizeObserver 不存在于 jsdom 中
  const MockObserverInstance: ResizeObserver = {
    disconnect: jest.fn(),
    unobserve: jest.fn(),
    observe: jest.fn(),
  };
  beforeEach(() => {
    global.ResizeObserver = jest
      .fn()
      .mockImplementation(() => MockObserverInstance);
  });

  test('attrs', () => {
    const onClick = jest.fn();
    const { container } = render(
      <Dropdown
        attrs={{ style: { background: '#fff' }, className: 'foo', onClick }}
        appendTo={null}
        options={[]}
        visible
      >
        <Button>foo bar</Button>
      </Dropdown>,
    );
    const balloon = container.querySelector('.t-word-balloon');

    expect(balloon).toHaveClass('foo');
    expect(balloon).toHaveStyle({ background: '#fff' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(balloon!);
    expect(onClick).toBeCalled();
  });

  test('basic', () => {
    const onSelect = jest.fn();
    const options: DropdownOptionsItem[] = [
      { label: '黄金蛋炒饭', key: '1' },
      { type: 'divider', key: 'd' },
      { label: '扬州炒饭', key: '2' },
    ];

    const { container } = render(
      <Dropdown onSelect={onSelect} options={options} appendTo={null} visible>
        <Button>foo bar</Button>
      </Dropdown>,
    );
    const balloon = container.querySelector('.t-word-balloon');
    expect(balloon).toMatchSnapshot();

    expect(onSelect).not.toBeCalled();
    fireEvent.click(balloon!.querySelector('.t-option')!);
    expect(balloon).toMatchSnapshot();
    expect(onSelect).toBeCalled();
    expect(onSelect.mock.calls[0]).toEqual([
      { label: '黄金蛋炒饭', key: '1' }, // option
      [], // parent
    ]);
    expect(balloon).toHaveClass('t-dropdown-leave-to');
  });

  test('disabled', () => {
    const onSelect = jest.fn();
    const options: DropdownOptionsItem[] = [
      { label: '黄金蛋炒饭', disabled: true, key: '1' },
      { label: '扬州炒饭', key: '2' },
    ];

    const { container } = render(
      <Dropdown onSelect={onSelect} options={options} appendTo={null} visible>
        <Button>foo bar</Button>
      </Dropdown>,
    );
    const balloon = container.querySelector('.t-word-balloon');
    expect(balloon).toMatchSnapshot();

    expect(onSelect).not.toBeCalled();
    fireEvent.click(balloon!.querySelector('.t-option')!);
    expect(onSelect).not.toBeCalled();
    expect(balloon).not.toHaveClass('t-dropdown-leave-to');
  });
  test('hideOnClick', () => {
    const onSelect = jest.fn();
    const options: DropdownOptionsItem[] = [
      { label: '黄金蛋炒饭', key: '1' },
      { label: '扬州炒饭', key: '2' },
    ];

    const { container } = render(
      <Dropdown
        hideOnClick={false}
        onSelect={onSelect}
        options={options}
        appendTo={null}
        visible
      >
        <Button>foo bar</Button>
      </Dropdown>,
    );
    const balloon = container.querySelector('.t-word-balloon');
    expect(onSelect).not.toBeCalled();
    fireEvent.click(balloon!.querySelector('.t-option')!);
    expect(onSelect).toBeCalled();
    expect(balloon).not.toHaveClass('t-dropdown-leave-to');
  });
  test('header footer', () => {
    const options: DropdownOptionsItem[] = [
      { label: '黄金蛋炒饭', key: '1' },
      { label: '扬州炒饭', key: '2' },
    ];

    const { container } = render(
      <Dropdown
        footer={<div>footer</div>}
        header={<h1>header</h1>}
        hideOnClick={false}
        options={options}
        appendTo={null}
        visible
      >
        <Button>foo bar</Button>
      </Dropdown>,
    );
    const balloon = container.querySelector('.t-word-balloon');
    expect(balloon).toMatchSnapshot();
  });

  test('group', () => {
    const options: DropdownOptionsItem[] = [
      {
        children: [
          {
            key: 'shreddedChicken',
            label: '手撕鸡',
          },
          {
            children: [
              {
                label: '黄金蛋炒饭',
                key: 'hjdcf',
              },
              {
                label: '扬州炒饭',
                key: 'yzcf',
              },
            ],
            key: 'friedRice',
            type: 'group',
            label: '炒饭',
          },
        ],
        type: 'group',
        label: '饭菜类',
        key: 'meals',
      },
      {
        key: 'others',
        label: '其他',
      },
    ];

    const { container } = render(
      <Dropdown options={options} appendTo={null} visible>
        <Button>foo bar</Button>
      </Dropdown>,
    );
    const balloon = container.querySelector('.t-word-balloon');
    expect(balloon).toMatchSnapshot();
  });
  test('nest', async () => {
    const onSelect = jest.fn();
    const options: DropdownOptionsItem[] = [
      { label: '手撕鸡', key: '1' },
      {
        children: [
          { label: '黄金蛋炒饭,黄金蛋炒饭', key: '4' },
          { label: '扬州炒饭', key: '5' },
        ],
        label: (
          <Tooltip title={'轮胎3🌟推荐'}>
            <div>蛋炒饭</div>
          </Tooltip>
        ),
        attrs: { className: 'nest-option' },
        key: '2',
      },
      { type: 'divider', key: 'd1' },
      { label: '榴莲', key: '3' },
    ];

    const { container } = render(
      <Dropdown onSelect={onSelect} options={options} appendTo={null} visible>
        <Button>foo bar</Button>
      </Dropdown>,
    );
    fireEvent.mouseEnter(document.querySelector('.nest-option')!);
    await act(() => nextTick());

    const balloon = container.querySelector('.t-word-balloon');
    expect(onSelect).not.toBeCalled();
    expect(balloon).toMatchSnapshot();
  });
  it('修复测试 在鼠标右击启动的情况下，点击选项关闭窗体后无法再次鼠标右击开启 #79', () => {
    jest.useFakeTimers();
    const options: DropdownOptionsItem[] = [
      { label: '黄金蛋炒饭', key: '1' },
      { label: '扬州炒饭', key: '2' },
    ];
    const { container } = render(
      <Dropdown trigger="contextmenu" options={options}>
        <div>foo bar</div>
      </Dropdown>,
    );

    expect(getBalloon()).toBeNull();

    // 第一步：开启
    fireEvent.contextMenu(container.firstChild!);
    act(() => jest.advanceTimersByTime(500));
    act(() => jest.advanceTimersByTime(500));
    expect(getBalloon()).not.toBeNull();
    expect(getBalloon()).not.toHaveClass('t-transition--invisible');

    // 第二步：点击选项
    fireEvent.click(document.querySelector('.t-dropdown-option')!);
    act(() => jest.advanceTimersByTime(500));
    act(() => jest.advanceTimersByTime(500));
    expect(getBalloon()).toHaveClass('t-transition--invisible');

    // 第三步：再次开启
    fireEvent.contextMenu(container.firstChild!);
    act(() => jest.advanceTimersByTime(500));
    act(() => jest.advanceTimersByTime(500));
    // bug 所在
    expect(getBalloon()).not.toHaveClass('t-transition--invisible');

    function getBalloon(): HTMLDivElement | null {
      return document.querySelector('.t-word-balloon');
    }
  });
});
