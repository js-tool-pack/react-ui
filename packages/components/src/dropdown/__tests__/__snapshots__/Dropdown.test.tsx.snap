// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown basic 1`] = `
<div
  class="t-word-balloon t-dropdown t-word-balloon--bottom-start t-word-balloon--no-arrow t-dropdown-enter-active t-dropdown-enter-to"
  style="top: 10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <div
      class="t-dropdown__body"
    >
      <ul
        class="t-dropdown__options"
      >
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            黄金蛋炒饭
          </div>
        </li>
        <li
          class="t-divider t-dropdown__divider t-divider--center t-divider--horizontal"
          style="--t-divider-border-style: solid; --t-divider-border-width: 1px;"
        />
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            扬州炒饭
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown basic 2`] = `
<div
  class="t-word-balloon t-dropdown t-word-balloon--bottom-start t-word-balloon--no-arrow t-dropdown-leave-active t-dropdown-leave-to"
  style="top: 10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <div
      class="t-dropdown__body"
    >
      <ul
        class="t-dropdown__options"
      >
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            黄金蛋炒饭
          </div>
        </li>
        <li
          class="t-divider t-dropdown__divider t-divider--center t-divider--horizontal"
          style="--t-divider-border-style: solid; --t-divider-border-width: 1px;"
        />
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            扬州炒饭
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown disabled 1`] = `
<div
  class="t-word-balloon t-dropdown t-word-balloon--bottom-start t-word-balloon--no-arrow t-dropdown-enter-active t-dropdown-enter-to"
  style="top: 10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <div
      class="t-dropdown__body"
    >
      <ul
        class="t-dropdown__options"
      >
        <li
          class="t-option t-dropdown-option t--size-m"
          disabled=""
          role="option"
        >
          <div
            class="t-option__label"
          >
            黄金蛋炒饭
          </div>
        </li>
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            扬州炒饭
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown group 1`] = `
<div
  class="t-word-balloon t-dropdown t-word-balloon--bottom-start t-word-balloon--no-arrow t-dropdown-enter-active t-dropdown-enter-to"
  style="top: 10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <div
      class="t-dropdown__body"
    >
      <ul
        class="t-dropdown__options"
      >
        <li
          class="t-dropdown__group"
        >
          <div
            class="t-option t-dropdown-option t-dropdown__group-title t--size-m t-option--readonly"
            role="heading"
          >
            <div
              class="t-option__label"
            >
              饭菜类
            </div>
          </div>
          <ul
            class="t-dropdown__group-body"
          >
            <li
              class="t-option t-dropdown-option t--size-m"
              role="option"
            >
              <div
                class="t-option__label"
              >
                手撕鸡
              </div>
            </li>
            <li
              class="t-dropdown__group"
            >
              <div
                class="t-option t-dropdown-option t-dropdown__group-title t--size-m t-option--readonly"
                role="heading"
              >
                <div
                  class="t-option__label"
                >
                  炒饭
                </div>
              </div>
              <ul
                class="t-dropdown__group-body"
              >
                <li
                  class="t-option t-dropdown-option t--size-m"
                  role="option"
                >
                  <div
                    class="t-option__label"
                  >
                    黄金蛋炒饭
                  </div>
                </li>
                <li
                  class="t-option t-dropdown-option t--size-m"
                  role="option"
                >
                  <div
                    class="t-option__label"
                  >
                    扬州炒饭
                  </div>
                </li>
              </ul>
            </li>
          </ul>
        </li>
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            其他
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown header footer 1`] = `
<div
  class="t-word-balloon t-dropdown t-word-balloon--bottom-start t-word-balloon--no-arrow t-dropdown-enter-active t-dropdown-enter-to"
  style="top: 10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <div
      class="t-dropdown__header"
    >
      <h1>
        header
      </h1>
    </div>
    <div
      class="t-dropdown__body"
    >
      <ul
        class="t-dropdown__options"
      >
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            黄金蛋炒饭
          </div>
        </li>
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            扬州炒饭
          </div>
        </li>
      </ul>
    </div>
    <div
      class="t-dropdown__footer"
    >
      <div>
        footer
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown nest 1`] = `
<div
  class="t-word-balloon t-dropdown t-word-balloon--bottom-start t-word-balloon--no-arrow t-dropdown-enter-active t-dropdown-enter-to"
  style="top: 10px; left: 0px;"
>
  <div
    class="t-word-balloon__content"
  >
    <div
      class="t-dropdown__body"
    >
      <ul
        class="t-dropdown__options"
      >
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            手撕鸡
          </div>
        </li>
        <li
          class="t-option t-dropdown-option nest-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            <div>
              蛋炒饭
            </div>
            <div
              class="t-word-balloon t-dropdown t-dropdown__nest t-word-balloon--right-start t-word-balloon--no-arrow t-dropdown-enter-active t-dropdown-enter-to"
              style="top: 0px; left: 0px;"
            >
              <div
                class="t-word-balloon__content"
              >
                <div
                  class="t-dropdown__body"
                >
                  <ul
                    class="t-dropdown__options"
                  >
                    <li
                      class="t-option t-dropdown-option t--size-m"
                      role="option"
                    >
                      <div
                        class="t-option__label"
                      >
                        黄金蛋炒饭,黄金蛋炒饭
                      </div>
                    </li>
                    <li
                      class="t-option t-dropdown-option t--size-m"
                      role="option"
                    >
                      <div
                        class="t-option__label"
                      >
                        扬州炒饭
                      </div>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            class="t-option__extra"
          >
            <i
              class="t-icon t-dropdown-option__expand"
              role="img"
            >
              <svg
                viewBox="0 0 1024 1024"
              >
                <path
                  d="M636.91389854 512.22982846L217.86080278 92.94690423l81.32081573-81.20590149 500.25899726 500.52713049L299.14331376 1012.25899726l-81.20590149-81.282511z"
                  fill="currentColor"
                />
              </svg>
            </i>
          </div>
        </li>
        <li
          class="t-divider t-dropdown__divider t-divider--center t-divider--horizontal"
          style="--t-divider-border-style: solid; --t-divider-border-width: 1px;"
        />
        <li
          class="t-option t-dropdown-option t--size-m"
          role="option"
        >
          <div
            class="t-option__label"
          >
            榴莲
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
